modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php

```diff
@@ -29,96 +31,165 @@ class AZParagraphsItem extends ParagraphsItem {
    * {@inheritdoc}
    */
   public function query() {
+    $query = parent::query();
+
     // @phpstan-ignore-next-line
     $this->allowArchivedParagraphs = \Drupal::config('az_migration.settings')->get('allow_archived_paragraphs');
 
-    $query = $this->select('paragraphs_item', 'p')
-      ->fields('p',
-        ['item_id',
-          'bundle',
-          'field_name',
-          'archived',
-          'bottom_spacing',
-          'view_mode',
-        ])
-      ->fields('pr', ['revision_id']);
-    $query->innerJoin('paragraphs_item_revision', 'pr', static::JOIN);
+    $query->addField('p', 'bottom_spacing', 'bottom_spacing');
+    $query->addField('p', 'view_mode', 'view_mode');
 
     if (!$this->allowArchivedParagraphs) {
-      // Omit archived (deleted or stale) paragraphs.
-      $query->condition('p.archived', 0);
-    }
-    // This configuration item may be set by a deriver to restrict the
-    // bundles retrieved.
-    if ($this->configuration['bundle']) {
-      $query->condition('p.bundle', $this->configuration['bundle']);
+      $this->excludeArchivedParagraphs($query);
     }
+
     return $query;
   }
 
   /**
    * {@inheritdoc}
    */
   public function fields() {
-    $fields = [
-      'item_id' => $this->t('The paragraph_item id'),
-      'revision_id' => $this->t('The paragraph_item revision id'),
-      'bundle' => $this->t('The paragraph bundle'),
-      'field_name' => $this->t('The paragraph field_name'),
+
+    return [
       'bottom_space' => $this->t('The paragraph Bottom Space'),
       'view_mode' => $this->t('The paragraph view mode'),
-    ];
-    return $fields;
+    ] + parent::fields();
+
   }
 
   /**
    * {@inheritdoc}
    */
   public function prepareRow(Row $row) {
-
-    // Get Item Id and revision Id of paragraph.
+    // Extract item_id and revision_id from the current row.
     $item_id = $row->getSourceProperty('item_id');
     $revision_id = $row->getSourceProperty('revision_id');
 
-    // Checking the field collection fields present in the paragraph.
+    // Process field collections associated with the paragraph.
+    $this->processFieldCollections($row, $item_id, $revision_id);
+
+    // Process individual fields of the paragraph.
+    $this->processParagraphFields($row, $item_id, $revision_id);
+
+    return $row;
+  }
+
+  /**
+   * Processes field collections within a paragraph.
+   *
+   * Field collections are groups of fields that can be reused across different
+   * entities.
+   * This function iterates over each field collection defined in the source
+   * property and processes them individually.
+   *
+   * @param \Drupal\migrate\Row $row
+   *   The current row being processed.
+   * @param string|int $item_id
+   *   The item ID of the paragraph.
+   * @param string|int $revision_id
+   *   The revision ID of the paragraph.
+   */
+  private function processFieldCollections(Row $row, $item_id, $revision_id) {
```

Comment:
I think we may want to make this protected so child classes can use it. Otherwise I think this method seems unnecessary to me.